examples/idempotent_producer_example/idempotent_producer_example.go

```diff
@@ -0,0 +1,178 @@
+	}
+
+	// For signalling termination from main to go-routine
+	termChan := make(chan bool, 1)
+	// For signalling that termination is done from go-routine to main
+	doneChan := make(chan bool)
+
+	// Go routine for serving the events channel for delivery reports and error events.
+	go func() {
+		doTerm := false
+		for !doTerm {
+			select {
+			case e := <-p.Events():
+				switch ev := e.(type) {
+				case *kafka.Message:
+					// Message delivery report
+					m := ev
+					if m.TopicPartition.Error != nil {
+						fmt.Printf("Delivery failed: %v\n", m.TopicPartition.Error)
+					} else {
+						fmt.Printf("Delivered message to topic %s [%d] at offset %v\n",
+							*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset)
+					}
+
+				case kafka.Error:
+					// Generic client instance-level errors, such as
+					// broker connection failures, authentication issues, etc.
+					//
+					// These errors should generally be considered informational
+					// as the underlying client will automatically try to
+					// recover from any errors encountered, the application
+					// does not need to take action on them.
+					//
+					// But with idempotence enabled, truly fatal errors can
+					// be raised when the idempotence guarantees can't be
+					// satisfied, these errors are identified by
+					// `e.IsFatal()`.
+
+					e := ev
+					if e.IsFatal() {
+						// Fatal error handling.
+						//
+						// When a fatal error is detected by the producer
+						// instance, it will emit kafka.Error event (with
+						// IsFatal()) set on the Events channel.
+						//
+						// Note:
+						//   After a fatal error has been raised, any
+						//   subsequent Produce*() calls will fail with
+						//   the original error code.
+						fmt.Printf("FATAL ERROR: %v: terminating\n", e)
+						run = false
+					} else {
+						fmt.Printf("Error: %v\n", e)
+					}
+
+				default:
+					fmt.Printf("Ignored event: %s\n", ev)
+				}
+
+			case <-termChan:
+				doTerm = true
+			}
+		}
+
+		close(doneChan)
+	}()
+
+	msgcnt := 0
+	for run == true {
+		value := fmt.Sprintf("Go Idempotent Producer example, message #%d", msgcnt)
+
+		// Produce message.
+		// This is an asynchronous call, on success it will only
+		// enqueue the message on the internal producer queue.
+		// The actual delivery attempts to the broker are handled
+		// by background threads.
+		// Per-message delivery reports are emitted on the Events() channel,
+		// see the go-routine above.
+		err = p.Produce(&kafka.Message{
+			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
+			Value:          []byte(value),
+		}, nil)
+
+		if err != nil {
+			fmt.Printf("Failed to produce message: %v\n", err)
+		}
+
+		msgcnt++
+
+		// Since fatal errors can't be triggered in practice,
+		// use the test API to trigger a fabricated error after some time.
+		if msgcnt == 13 {
+			p.TestFatalError(kafka.ErrOutOfOrderSequenceNumber, "Testing fatal errors")
```

Comment:
Although you've convinced me of the value of exposing this test method, I don't think it should be on the producer instance. My first instinct is to have a new test class with a method 'CauseFatalError' on it which takes a producer instance. A side effect of this (but not the primary driver, which is structural) is I also like having a different verb in there 'cause' or 'emit', since test could be considered an adjective. but we definitely don't want to drop the word test.

Reply:
There are no classes in Go, the method needs to be super generic or be a Producer-typed interface - since this method is only relevant for the producer I think it resides in the right place.

I think Test should be in there, which is a verb in this case, so I dont think an additional verb (Cause) adds anything.
Anything named Test should stand out anyway and will not be confused with production code.